src/main/scala/com/fulcrumgenomics/commons/async/AsyncRunnable.scala

```diff
@@ -0,0 +1,168 @@
+  *
+  * The `checkAndRaise()` method should be used by implementing classes to check if the
+  * `execute()`  method has encountered an exception, and if so, it will rethrow the exception as an
+  * [[Error]] or [[RuntimeException]].
+  *
+  * The `tryAndModifyInterruptedException()` method should wrap blocking code in implementing classes
+  * that are not part of the `execute()` method.  This method will throw a [[RuntimeException]] if interrupted while
+  * waiting.
+  */
+trait AsyncRunnable extends Runnable {
+
+  /** A reference to an exception thrown during an asynchronous operation. */
+  private[AsyncRunnable] val _throwable: AtomicReference[Throwable] = new AtomicReference(null)
+
+  /** This signals that the run method has started. */
+  private val startedLatch: CountDownLatch = new CountDownLatch(1)
+
+  /** This signals that the run method has completed, even if an exception occurred. This is needed so the close method
+    * can be sure that the run method is no longer executing. */
+  private val doneLatch: CountDownLatch = new CountDownLatch(1)
+
+  final def run(): Unit = {
+    val outer = this
+    startedLatch.countDown()
+    try {
+      outer.execute()
+    } catch {
+      case thr: Throwable =>
+        outer._throwable.compareAndSet(null, thr)
+        outer.uponException()
+    } finally {
+      outer.uponFinally()
+      this.doneLatch.countDown()
+    }
+  }
+
+  /** Returns true if the [[run()]] method has started, false otherwise. */
+  def started: Boolean = this.startedLatch.getCount == 0
+
+  /** Returns true if the [[run()]] method has completed, false otherwise. */
+  def done: Boolean = this.doneLatch.getCount == 0
+
+  /** Returns a throwable if an exception occurred in the [[run()]] method, None otherwise. */
+  def throwable: Option[Throwable] = Option(this._throwable.get())
+
+  /** Waits for the [[run()]] method to start. */
+  def awaitStart(): Unit = this.startedLatch.await()
+
+  /** Returns an [[Awaitable]] that completes when the [[run()]] method has started.  Returns the throwable if an
+    * exception was encountered, [[None]] otherwise. */
+  def startAwaitable()(implicit ec: ExecutionContext) : Awaitable[Unit] = Future { this.awaitStart() }
+
+  /** Waits for the [[run()]] method to complete. */
+  def awaitDone(): Unit = this.doneLatch.await()
+
+  /** Returns an [[Awaitable]] that completes when the [[run()]] method has completed.  Returns the throwable if an
+    * exception was encountered, [[None]] otherwise. */
+  def doneAwaitable()(implicit ec: ExecutionContext) : Awaitable[Option[Throwable]] = Future { this.awaitDone(); Option(this._throwable.get()) }
+
+  /** Creates a new thread wrapping this runnable; the thread is not started.
+    *
+    * @param name optionally the name of the thread, otherwise a name is created based on the class name and the number
+    *             of threads already created.
+    */
+  final def thread(name: Option[String] = None): Thread = {
```

Comment:
I'm curious why this is final and public?  I.e. I can't override it, and should I really be calling it directly?

Reply:
This method gives you control of what to do with the thread without it being started.  It's final, since I don't want folks messing with this method, as it has some logic that isn't public about setting the thread name.  We can loosen the `final` later if we want to. I  was just erring on the side of saying "hey it's final because there's some stuff in here that you can't access if you override it".  ¯\_(ツ)_/¯